include/ctpg/ctpg.hpp

```diff
@@ -1638,7 +1644,61 @@ namespace detail
     constexpr size_t value_stack_initial_capacity = 1 << 10;
     constexpr size_t cursor_stack_initial_capacity = 1 << 10;
 
-    template<typename CursorStack, typename ValueStack, typename ErrorStream, typename Iterator>
+    template<typename Context, typename ValueVariantType, typename RuleTupleType, size_t RuleCount>
+    struct value_reductors
+    {
+        constexpr value_reductors(Context &&context, const RuleTupleType& rule_tuple): context(context), rule_tuple(rule_tuple) {
+            init_reductors(rule_tuple, std::make_index_sequence<std::tuple_size_v<RuleTupleType>>{});
+        }
+
+        constexpr ValueVariantType invoke(size_t i, ValueVariantType* args) const {
+            return reductors[i](context, rule_tuple, args);
+        }
+
+    private:
+        template<size_t... I>
+        constexpr void init_reductors(const RuleTupleType& rule_tuple, std::index_sequence<I...>)
+        {
+            (void(init_nth_reductor<I>(std::get<I>(rule_tuple))), ...);
+        }
+
+        template<size_t Nr, bool RequiresContext, typename F, typename L, typename... R>
+        constexpr void init_nth_reductor(const detail::rule<RequiresContext, F, L, R...>&)
+        {
+            reductors[Nr] = &reduce_value<Nr, RequiresContext, F, value_type_t<L>, value_type_t<R>...>;
+        }
+
+        template<bool RequiresContext, typename F, typename LValueType, typename... RValueType, size_t... I>
+        constexpr static LValueType reduce_value_impl([[maybe_unused]] Context& ctx, [[maybe_unused]] const F& f, ValueVariantType* start, std::index_sequence<I...>)
+        {
+            if constexpr (std::is_same_v<F, std::nullptr_t>)
+                return LValueType(std::get<RValueType>(std::move(*(start + I)))...);
+            else
+            {
+                if constexpr (RequiresContext)
+                    return LValueType(f(ctx, std::get<RValueType>(std::move(*(start + I)))...));
+                else
+                    return LValueType(f(std::get<RValueType>(std::move(*(start + I)))...));
+            }
+        }
+
+        template<size_t RuleIdx, bool RequiresContext, typename F, typename LValueType, typename... RValueType>
+        constexpr static ValueVariantType reduce_value(Context& ctx, const RuleTupleType& rules, ValueVariantType* start)
+        {
+            return ValueVariantType(
+                reduce_value_impl<RequiresContext, F, LValueType, RValueType...>(
+                    ctx, std::get<RuleIdx>(rules).get_f(), start, std::index_sequence_for<RValueType...>{})
+            );
+        }
+
+        Context &context;
```

Comment:
Is it always referrence? What if I pass context as a temporary object?

Reply:
Then `Context&` should be deduced as const reference, I suppose. But I'm a bit confused with storing forwarded references as fields, so, I may be wrong.

What is the proper way to store a reference that can be either const and non-const?

Also, instead of storing it, I can forward it to `invoke` from `reduce`.

Reply:
If we can avoid stoing it, it wuld the best.